### DISCUSSION TITLE
Where are the C++ kernel implementation of the Winograd convolution algorithm?

### DISCUSSION THREAD

**Discussion by Redtorm** [ROCmSoftwarePlatform/MIOpen#1686] [Q&A]
I would like to know how MIOpen implements the Winograd convolution algorithm, but I only found the convolution kernel implemented in assembly language in the project source code, which is difficult to read and understand. I guess these assembly kernels should be converted from C++ kernels, but I can't find where the related C++ kernel implementations are, **could you tell me where these C++ kernel implementation files are?** 

How can I add my own kernel to MIOpen? Is there a manual for this?

**Comment by junliume** [comment]
@Kirpich30000  could you comment on the question?

**Comment by Kirpich30000** [comment]
There is no C++ implementation - kernel were specifically designed for high efficiency and implemented from scratch in assembly. 
You could look at [this paper](https://arxiv.org/abs/1509.09308) for Winograd introduction. And here is a [reference code](https://github.com/andravin/wincnn) for constructing matrices if you want to try customized Winograd transformation.

**Comment by Kirpich30000** [comment]
To add your own kernel you will need to add a solver and register it in [solver.cpp](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/solver.cpp). Peek any solver and look at the code to see how it is implemented. Some solvers use workspace, others - don't. Some solvers invokes only a single kernel, while the other could invoke multiple kernels to perform the operation.

**Comment by atamazov** [comment]
@junliume I recommend removing `request_for_comments` (this is `question`, RFC is more like a [spec-in-progress](https://en.wikipedia.org/wiki/Request_for_Comments)) and, maybe, closing this.

**Comment by DrizztDoUrden** [comment]
Also to register the solver for the specific convolution (or not conv) calls it supports like these: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/d19c2575ca918f6b63a022f004d591795afcd85d/src/mlo_dir_conv.cpp#L110

**Comment by atamazov** [comment]
Intro info about solvers: #866
